nixos/tests/dsl.nix

```diff
@@ -0,0 +1,247 @@
+    };
+  Node = (
+    { name, ... }:
+    {
+      options.name = lib.mkOption {
+        type = lib.types.str;
+        default = name;
+      };
+      options.request = lib.mkEnableOption "Whether to mark the node for building";
+      options.needed = lib.mkOption {
+        type = with lib.types; nullOr bool;
+        default = null;
+        description = "Whether the node must be built to satisfy all requests (either a requested node or a transitive dependency)";
+      };
+      options.chosen = lib.mkOption {
+        type = with lib.types; nullOr bool;
+        default = null;
+        description = "Whether the node is included in the build plan (i.t. it's `needed` and fitted into budget)";
+      };
+      options.cache = lib.mkOption {
+        type = lib.types.enum [
+          "none"
+          "remote"
+          "local"
+        ];
+        description = ''
+          Whether the dependency is pre-built and available in the local /nix/store ("local"), can be substituted ("remote"), or has to be built ("none")
+        '';
+        default = "none";
+      };
+      options.inputs = lib.mkOption {
+        type = lib.types.attrsOf (lib.types.submodule Dependency);
+        default = { };
+      };
+    }
+  );
+  Nodes = lib.types.attrsOf (lib.types.submodule Node);
+  scope-fun = import ./scope-fun.nix {
+    inherit lib;
+    inherit (config.dag) nodes;
+  };
+in
+{
+  options.dag = {
+    nodes = lib.mkOption {
+      type = Nodes;
+      description = "Derivation DAG, including cache status and references.";
+    };
+    needBuilds = lib.mkOption {
+      type = with lib.types; nullOr int;
+      default = null;
+      description = "How many builds are required to satisfy all targets. Null disables the test";
+    };
+    needDownloads = lib.mkOption {
+      type = with lib.types; nullOr int;
+      default = null;
+      description = "How many downloads are required to satisfy all targets. Null disables the test";
+    };
+    choseBuilds = lib.mkOption {
+      type = with lib.types; nullOr int;
+      default = null;
+      description = "How many builds we expect evanix to choose to satisfy the maximum number of targets within the given budget. Null disables the test";
+    };
+    choseDownloads = lib.mkOption {
+      type = with lib.types; nullOr int;
+      default = null;
+      description = "How many downloads we expect evanix to choose to satisfy the maximum number of targets within the given budget. Null disables the test";
+    };
+    allowBuilds = lib.mkOption {
+      type = with lib.types; nullOr int;
+      default = null;
+      description = "How many builds evanix is allowed to choose. Null means no constraint";
+    };
+    allowDownloads = lib.mkOption {
+      type = with lib.types; nullOr int;
+      default = null;
+      description = "How many downloads evanix is allowed to choose. Null means no constraint";
```

Comment:
The first one is to be passed to evanix as `--max-builds` option. We could add more constraint options like `--max-substitutions`, `--max-time`. Then we describe the solution we expect using `nodes.<name>.chosen`, and the test driver ensures that what evanix outputs matches the expectation (not implemented)